### PR TITLE
Support Level 1 dot/dotc/dotu

### PR DESCRIPTION
**Pull request by kballeda**

### CODE REVIEW

**Comment by kballeda** [comment]
l1_dot_pass

**Comment by kballeda** [comment]
@pengtu / @Sarbojit2019  - Please review this PR.

**Review by pengtu — APPROVED** [review]
Looks great!  Thanks.

File: test/onemkl.jl [hunk 1]
```diff
@@ -62,8 +62,17 @@ m = 20
...
         end
 
+        @testset "dot" begin
+            @test testf(dot, rand(T,m), rand(T,m))
+            if T == ComplexF32 || T == ComplexF64
+                @test testf(oneMKL.dotu, m, 
+                            oneArray(rand(T,m)),
+                            oneArray(rand(T,m)))
+            end
+
```
Comment (Sarbojit2019) [inline test/onemkl.jl:73]: 'end' is missed??

File: lib/mkl/libonemkl.jl [hunk 2]
```diff
@@ -42,6 +42,44 @@ function onemklZgemm(device_queue, transA, transB, m, n, k, alpha, A, lda, B, ld
...
+                                        incy::Int64, result::RefOrZeRef{ComplexF32})::Cvoid
+end
+
+function onemklZdotu(device_queue, n, x, incx, y, incy, result)
+    @ccall liboneapi_support.onemklZdotu(device_queue::syclQueue_t, n::Int64,
+                                        x::ZePtr{ComplexF64}, incx::Int64, y::ZePtr{ComplexF64},
+                                        incy::Int64, result::RefOrZeRef{ComplexF64})::Cvoid
+end
+
+
```
Comment (Sarbojit2019) [inline lib/mkl/libonemkl.jl]: can we remove extra newline here.

File: lib/mkl/libonemkl.jl [hunk 3]
```diff
@@ -42,6 +42,44 @@ function onemklZgemm(device_queue, transA, transB, m, n, k, alpha, A, lda, B, ld
                                     C::ZePtr{ComplexF64}, ldc::Int64)::Cint
 end
 
+
```
Comment (Sarbojit2019) [inline lib/mkl/libonemkl.jl]: extra new line

File: deps/src/onemkl.cpp [hunk 4]
```diff
@@ -87,6 +87,64 @@ extern "C" int onemklZgemm(syclQueue_t device_queue, onemklTranspose transA,
...
+                           const double _Complex *x, int64_t incx, const double _Complex *y,
+                           int64_t incy, double _Complex *result) {
+    auto status = oneapi::mkl::blas::column_major::dotu(device_queue->val, n,
+                                                reinterpret_cast<const std::complex<double> *>(x), incx,
+                                                reinterpret_cast<const std::complex<double> *>(y), incy,
+                                                reinterpret_cast<std::complex<double> *>(result));
+    __FORCE_MKL_FLUSH__(status);
+}
+
+
```
Comment (Sarbojit2019) [inline deps/src/onemkl.cpp]: extra new line

**Review by Sarbojit2019 — COMMENTED** [review]
Mostly cosmetic changes, would be good to maintain sanity.

File: lib/mkl/libonemkl.jl (hunk 2 above)
Comment (kballeda) [reply lib/mkl/libonemkl.jl]: Done

File: lib/mkl/libonemkl.jl (hunk 3 above)
Comment (kballeda) [reply lib/mkl/libonemkl.jl]: Done

File: deps/src/onemkl.cpp (hunk 4 above)
Comment (kballeda) [reply deps/src/onemkl.cpp]: Done

File: test/onemkl.jl (hunk 1 above)
Comment (kballeda) [reply test/onemkl.jl:73]: Good catch, yes looks like a merge issue.

**Comment by kballeda** [comment]
> Looks great! Thanks.

Thanks for the review.

**Comment by kballeda** [comment]
> Mostly cosmetic changes would be good to maintain sanity.

Thanks for the review.

**Comment by kballeda** [comment]
image

Result of CI run latest.

**Review by Sarbojit2019 — APPROVED** [review]

**Pull request merged**